src/bwt_util.rs

```diff
@@ -1,3 +1,111 @@
+/// resolution
+pub fn pairwise_merge_iter<T: Ord + Hash + Clone + Debug>(
+    current_interleave: &BitVec<u64, Msb0>,
+    bwt0: &[T],
+    bwt1: &[T],
+    offsets: &HashMap<T, usize>,
+) -> BitVec<u64, Msb0> {
+    // initial conditions
+    let mut next_interleave = bitvec![u64, Msb0; 0; current_interleave.len()];
+    let mut current_pos0 = 0usize;
+    let mut current_pos1 = 0usize;
+
+    let mut temp_index = offsets.clone();
+
+    // iterate through each bit
+    for b in current_interleave.iter() {
+        let c = if *b {
+            let v = bwt0[current_pos0].clone();
+            current_pos0 += 1;
+            v
+        } else {
+            let v = bwt1[current_pos1].clone();
+            current_pos1 += 1;
+            v
+        };
+        // copy b into the next position for c
+        let next_position = temp_index
+            .get_mut(&c)
+            .expect("Unknown character encountered in merging BWTs");
+        next_interleave.set(*next_position, *b);
+        // update the tempIndex to match the FM-index
+        *next_position += 1;
+    }
+
+    next_interleave
+}
+
+/// Determines the offset hashmap for a provided slice of BWTs.
+/// The offset hashmap is the mapping from a character to the number of
+/// lexigraphically lesser characters present in the combined string provided.
+fn generate_offset_hashmap<T: Ord + Hash + Clone + Debug>(bwts: &[&[T]]) -> HashMap<T, usize> {
+    let mut num_occurrences: HashMap<T, usize> = HashMap::new();
+    for &bwt in bwts.iter() {
+        for c in bwt {
+            if let Some(i) = num_occurrences.get_mut(c) {
+                *i += 1;
+            } else {
+                num_occurrences.insert(c.clone(), 1);
+            }
+        }
+    }
+    let ordered_chars = num_occurrences.keys().sorted().cloned().collect::<Vec<T>>();
+    let mut total = 0usize;
+    let mut offset_map = HashMap::with_capacity(num_occurrences.len());
+    for c in ordered_chars {
+        offset_map.insert(c.clone(), total);
+        total += num_occurrences[&c];
+    }
+    offset_map
+}
+
+/// Merges two Burrows-Wheeler transformed slices into a new, merged Burrows-Wheeler
+/// slice.
+pub fn pairwise_bwt_merge<T: Ord + Hash + Clone + Debug>(bwt0: &[T], bwt1: &[T]) -> Vec<T> {
+    let total_len = bwt0.len() + bwt1.len();
+    let initial_offsets = generate_offset_hashmap(&[bwt0, bwt1]);
+
+    let mut interleave = bitvec![u64, Msb0; 0; total_len];
+    // initialize ret array to 0s followed by 1s
+    let mut final_interleave = bitvec![u64, Msb0; 0; total_len];
+    for i in bwt0.len()..total_len {
+        // TODO - figure out if there's a way to do this faster than iterating through each element
+        // and setting them one by one. (BitVec should have option to do a bitwise OR)
+        final_interleave.set(i, true)
+    }
+    while interleave != final_interleave {
+        // copy the old interleaving and re-iterate
+        interleave = final_interleave.clone();
```

Comment:
Pretty sure you can remove the `.clone()` part of this statement entirely and save some mem-copy performance on large datasets.